src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/FrameConnection.cs

```diff
@@ -231,21 +255,49 @@ private void DisposeAdaptedConnections()
 
         public void Tick(DateTimeOffset now)
         {
-            Debug.Assert(_frame != null, $"nameof({_frame}) is null");
+            Debug.Assert(_frame != null, $"{nameof(_frame)} is null");
 
             var timestamp = now.Ticks;
 
-            // TODO: Use PlatformApis.VolatileRead equivalent again
-            if (timestamp > Interlocked.Read(ref _timeoutTimestamp))
+            if (timestamp > Interlocked.Read(ref _timeoutTimestamp)) // TODO: Use PlatformApis.VolatileRead equivalent again
             {
                 CancelTimeout();
 
                 if (_timeoutAction == TimeoutAction.SendTimeoutResponse)
                 {
+                    SetTimeoutResponse();
+                }
+
+                Timeout();
+            }
+            else if (_timingReads)
+            {
+                _readTimingElapsed += timestamp - _lastTimestamp;
+
+                if (_readTimingElapsed > _frame.RequestBodyTimeout.Ticks)
+                {
+                    Log.RequestBodyTimeout(_context.ConnectionId, _frame.TraceIdentifier, _frame.RequestBodyTimeout);
                     Timeout();
                 }
+                else if (_frame.RequestBodyMinimumDataRate > 0)
+                {
+                    var rate = (_bytesRead / (_readTimingElapsed / TimeSpan.TicksPerSecond));
+
+                    if (_readTimingElapsed > _frame.RequestBodyMinimumDataRateGracePeriod.Ticks && rate < _frame.RequestBodyMinimumDataRate)
+                    {
+                        Log.RequestBodyMininumDataRateNotSatisfied(_context.ConnectionId, _frame.TraceIdentifier, _frame.RequestBodyMinimumDataRate);
+                        Timeout();
+                    }
+                }
 
-                _frame.Stop();
+                // PauseTimingReads() cannot just set _timingReads to false. It needs to go through at least one tick
+                // before pausing, otherwise _readTimingElapsed might never be updated if PauseTimingReads() is always
+                // called before the next tick.
```

Comment:
This means that elapsed time is being added while there is no read callback configured. This means about a second of elapsed time could be added for each pause when the client actually can't upload additional data.

If we error, we should error the other way.